aten/src/ATen/cuda/CUDAApplyUtils.cuh

```diff
@@ -955,5 +956,199 @@ bool CUDA_tensor_apply4(at::Tensor a,
+      }
+    }
+  }
+}
+
+#define MAX_NUMBER_BINS_WITH_SHARED_MEM 5000
+
+/*
+  Calculate the frequesncy of the input values.
+
+  `a` contains the final output or the histogram. Input `b` is assumed to
+  be 1-D non-negative int array. `c` optionally contains the weight vector.
+  See `help torch.bincount` for details on the math.
+
+  3 implementations based of input size and memory usage:
+    MULTI_BLOCK: `partial_output` temp tensor with (#blocks X #bins) shape.
+        Each block atomically adds to it's own **global** hist copy:
+        `partial_output[i, :]`. Once all blocks are done, the partial output is
+        sum-reduced to give the final output with (#bins) shape.
+    SHARED: `partial_output` temp tensor with (#blocks X #bins) shape.
+        Each block atomically adds to it's own **shared** hist copy: `smem[:]`.
+        Copies it's local histogram to `partial_output[i, :]` followed by
+        sum-reduction to yield the final output.
+    GLOBAL: if #bins > MAX_NUMBER_BINS_WITH_SHARED_MEM. In this case, only a
+        single **global** copy of the output histogram is maintained and all
+        blocks/threads atomically update bin counts. Hence, this can be slow.
+        CPU version may perform better.
+ */
+template <typename scalar1, typename scalar2>
+bool CUDA_tensor_histogram(
+    at::Tensor a, /* output */
+    at::Tensor b, /* input */
+    at::Tensor c, /* weights(optional) */
+    int64_t nbins,
+    int binsize,
+    TensorArgType aType = TensorArgType::ReadWrite,
+    TensorArgType bType = TensorArgType::ReadOnly,
+    TensorArgType cType = TensorArgType::ReadOnly) {
+  checkBackend("CUDA_tensor_histogram", {a, b}, Backend::CUDA);
+  constexpr int has_weights = !std::is_integral<scalar1>::value;
+  if (has_weights) {
+    checkBackend("CUDA_tensor_histogram", {c}, Backend::CUDA);
+  }
+  int64_t totalElements = b.numel();
+
+  const dim3 block = getApplyBlock();
+  dim3 grid;
+  if (!getApplyGrid(totalElements, grid)) {
+    return false;
+  }
+#if CUDA_VERSION < 9000
+  grid.x = std::min(
+      (unsigned int)at::globalContext()
+              .getCurrentDeviceProperties()
+              ->multiProcessorCount *
+          AT_APPLY_BLOCKS_PER_SM,
+      grid.x);
+#endif
+
+  Tensor partial_output;
+  if (nbins < MAX_NUMBER_BINS_WITH_SHARED_MEM) {
+    partial_output = a.type().zeros({grid.x, nbins});
+  } else {
+    partial_output = a.type().zeros({1, nbins});
+  }
+  // kernel call to compute partial histogram
+  using IndexType = uint64_t;
+  auto pInfo = detail::getTensorInfo<scalar1, IndexType>(partial_output);
+  auto bInfo = detail::getTensorInfo<scalar2, IndexType>(b);
+  auto maxSharedMem =
+      at::globalContext().getCurrentDeviceProperties()->sharedMemPerBlock;
+  auto sharedMem = nbins * sizeof(scalar1) + 8; // 8 guard bytes
+
+#define GET_WEIGHTS_DUMMY_OP [] __device__(IndexType) { return 1L; }
+#define GET_WEIGHTS_OP                                                    \
```

Comment:
Are the macros necessary? It would be nice to make them static functions so the code is more maintainable